### DISCUSSION TITLE
Read and Write simultaneously with same database

### DISCUSSION THREAD

**Discussion by msarkrish** [simolus3/drift#2729] [Q&A]
Hi @simolus3 ,
Our application is developed with offline support. So once the user is logged into our application, we will fetch the necessary data for the logged-in user and store it in the local database. Currently, what we are doing is: After a successful login, we are displaying the loading screen until the data download and insert are complete. For some users, it is taking more than 10 minutes. So currently, what we are thinking is that after logging in, we can allow the user to navigate to the home screen and allow the user to use the app. In the background, we will insert the data. 

My questions are,
1. Is it possible to insert and write to the same table and database at the same time?
2. We are accessing the database from the native Android and iOS sides. Whether native access will affect isolate?
3. We are using sqlchiper. Does that also work with isolate?
4. What is the best way to handle large number of data, In our application some tables will have more than 4 Lac data and UI also watch the query streams to display the updated data. Is it good to use drift isolate or default drift createInBackground is enough for that?

**Comment by simolus3** [comment]
> For some users, it is taking more than 10 minutes

The insert alone or both the download and the insert? If you split the inserts the insert into smaller batches, you can absolutely use reads and writes concurrently. If you want the initial insert to happen entirely on a background isolate, you could also use [write-ahead logging](https://www.sqlite.org/wal.html) and open the database on multiple isolates - that way, the reads and writes are actually happening in parallel.

> Whether native access will affect isolate?

In this case I recommend enabling WAL even more, it decreases the chance of readers and writes blocking each other. You can also set the [busy timeout pragma](https://www.sqlite.org/pragma.html#pragma_busy_timeout) to a few milliseconds to have sqlite3 retry operations automatically.

> We are using sqlchiper. Does that also work with isolate?

This shouldn't change anything.

> Is it good to use drift isolate or default drift createInBackground is enough for that?

Using `createInBackground` is enough for almost all use-cases. If you have heavy tasks that need to do lots of writes, offloading the entire task into another isolate and giving it its own database connection may speed things up. But this is rarely useful for reads.

> UI also watch the query streams to display the updated data

If you have a stream with thousands of rows that is updating very frequently, that might be a source of slowdowns. If you run into problems with that, you could try using `limit` to only show the first 100 or so rows and deal with pagination by adding an offset to the limit.
